proto/contract.proto

```diff
@@ -11,14 +11,29 @@ option java_outer_classname = "DataPlaneContract";
 // Protobuf include nightmare? No thanks!
 message Empty {}
 
+// A dialect for Cloud Events Subscriptions API Filters
+enum Dialect {
+  EXACT = 0;
+  PREFIX = 1;
+  SUFFIX = 2;
+  ALL = 3;
+  ANY = 4;
+  NOT = 5;
+  CESQL = 6;
+}
+
 message Filter {
```

Comment:
Can't we have a `oneof` and then have concrete types for each of the dialects?

Since, we can only have one of the dialects set?

Something like:

```proto
message Filter {
  oneof filter {
    Exact exact = 1;
    Prefix prefix = 2;
    Suffix suffix = 3;
    All all = 4;
    Any any = 5;
    Not not = 6;
    Cesql cesql = 7;
  }

  message Exact {
    Attributes attributes = 1;
  }

  ...

  message All {
    Attributes attributes = 1;
    repeated Filter filters = 2;
  }

  ...
  
  message Cesql {
    string expression = 1;
  }

  message Attributes {
    map<string, string> attributes = 1;
  }
}
```

Not sure what's to be in `ALL` for sure but I think you would get the idea.

Reply:
Or, some mix of ENUM and the separate type could be better. Because for example, the contract definitions of `EXACT` and `PREFIX` are pretty much the same except how they're interpreted by the runtime.

Reply:
I'll check it out, sgtm.

Reply:
Note: nesting adds more space overhead and `oneof` is kinda hard to evolve.

In the past, I used this program as the base to see the space overhead of contract changes: https://gist.github.com/pierDipi/f2d6775c68d4c6b44b6fbfb6e73050e3

Reply:
> Note: nesting adds more space overhead.

I think it should be ok to use more space if we think this is a better structure.
Every nested thing means additional 2 bytes.

But, I know what you mean... I was doing some calc back then: https://github.com/aliok/catalli-public-code/blob/master/catalli-proto/protocol.proto#L47-L48

In this context, I don't think it will matter that much.

> and oneof is kinda hard to evolve.

Yeah it has some issues around backwards compatibility, especially when you delete a `oneof` field. I don't really expect us to delete fields. (adding new fields has its own problem but that would be fine as we are updating both writers and readers of the contract at the same time.)

I think we need to find the balance...